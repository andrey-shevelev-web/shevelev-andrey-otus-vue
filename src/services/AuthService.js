import Cookies from 'js-cookie';

const TOKEN_PARAM_NAME = 'main';
const TOKEN_EXPIRES = 1;

function setToken(token) {
  Cookies.set(TOKEN_PARAM_NAME, token, { expires: TOKEN_EXPIRES });
}

function getToken() {
  return Cookies.get(TOKEN_PARAM_NAME);
}

function removeToken() {
  Cookies.remove(TOKEN_PARAM_NAME);
}

export function authSignin(username, password) {
  let credentials = [username, password].join(':'),
    accessToken = `Basic ${btoa(credentials)}`;

  setToken(accessToken);
}

export function isUserAuthenticated() {
  return !!getToken();
}

export function authSignout() {
  removeToken();
}
