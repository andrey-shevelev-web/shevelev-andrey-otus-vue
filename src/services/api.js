import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'https://fakestoreapi.com',
  timeout: 180_000
});

export default axiosInstance;
