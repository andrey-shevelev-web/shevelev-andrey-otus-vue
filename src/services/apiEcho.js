import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'https://httpbin.org',
  timeout: 180_000
});

export default axiosInstance;
