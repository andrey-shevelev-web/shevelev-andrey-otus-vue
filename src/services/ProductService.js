import { jsonProducts } from '@/services/JsonProducts';
import api from '@/services/api';
import apiEcho from '@/services/apiEcho';

export function getProductsFromJson() {
  try {
    return JSON.parse(jsonProducts);
  } catch (error) {
    return {};
  }
}

export async function getProducts() {
  const response = await api.get('/products');
  return response.data;
}

export async function postOrderForm(formData) {
  const response = await apiEcho.post('/post', {
    params: { formData }
  });
  return response.data;
}
