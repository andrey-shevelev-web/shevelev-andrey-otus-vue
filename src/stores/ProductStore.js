import { ref, reactive, computed } from 'vue';
import { defineStore } from 'pinia';
import { v4 as uuidv4 } from 'uuid';
import { getProducts } from '@/services/ProductService';

export const useProductStore = defineStore('productStore', () => {
  const products = reactive([]);
  const isRequestRunning = ref(false);
  const requestError = ref('');
  const isVisibleCreateForm = ref(false);
  const productsFilter = ref('');

  const clearProducts = () => {
    products.splice(0, products.length);
  };

  const loadProducts = async () => {
    requestError.value = '';
    isRequestRunning.value = true;
    try {
      const responseData = await getProducts();
      clearProducts();
      products.push(...responseData);
    } catch (err) {
      requestError.value = 'Error! There was an error loading products';
    } finally {
      isRequestRunning.value = false;
    }
  };

  const setProductsFilter = value => {
    console.log('store value', value);
    productsFilter.value = value;
  };

  const filteredProducts = computed(() => {
    const rawFilter = productsFilter.value;

    if (rawFilter) {
      const priceFilter = Number(rawFilter);

      return products.filter(
        o =>
          o.title.includes(rawFilter) ||
          (!isNaN(priceFilter) && o.price === priceFilter)
      );
    } else {
      return products;
    }
  });

  const setVisibleCreateForm = value => {
    isVisibleCreateForm.value = value;
  };

  const createProduct = (title, category, description, price) => {
    const newProduct = {
      id: uuidv4(),
      title,
      category,
      description,
      price,
      image:
        'https://mir-baget.com/upload/iblock/b9b/nibxd6dvjqm4yv24tzzsvvv6y7w1rhzd.png',
      rating: {
        rate: 0,
        count: 0
      }
    };
    products.unshift(newProduct);
    setVisibleCreateForm(false);
  };

  return {
    products,
    loadProducts,
    productsFilter,
    setProductsFilter,
    filteredProducts,
    isRequestRunning,
    requestError,
    createProduct,
    isVisibleCreateForm,
    setVisibleCreateForm
  };
});
