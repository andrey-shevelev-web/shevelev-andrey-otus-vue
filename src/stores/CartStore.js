import { reactive, computed } from 'vue';
import { defineStore } from 'pinia';
import { v4 as uuidv4 } from 'uuid';
import { CART_DEC_QUANTITY } from '@/constants';
import { useProductStore } from '@/stores/ProductStore';

export const useCartStore = defineStore('cartStore', () => {
  const productStore = useProductStore();

  const cart = reactive([]);

  const clearCart = () => {
    cart.splice(0, cart.length);
  };

  const findCartItemByProductId = productId => {
    return cart.find(o => o.productId == productId);
  };

  const addProductToCart = productId => {
    const foundCartItem = findCartItemByProductId(productId);
    if (foundCartItem) {
      foundCartItem.quantity += 1;
    } else {
      const foundProduct = productStore.products.find(o => o.id == productId);
      if (!foundProduct) return;

      const newCartItem = {
        id: uuidv4(),
        productId,
        image: foundProduct.image,
        title: foundProduct.title,
        price: foundProduct.price,
        quantity: 1
      };

      cart.push(newCartItem);
    }
  };

  const changeCartItemQuantity = (productId, action) => {
    const foundCartItem = cart.find(o => o.productId == productId);

    if (foundCartItem) {
      switch (action) {
        case CART_DEC_QUANTITY:
          if (foundCartItem.quantity > 1) foundCartItem.quantity -= 1;
          break;
        default:
          foundCartItem.quantity += 1;
      }
    }
  };

  const removeCartItem = productId => {
    const foundCartItemIdx = cart.findIndex(o => o.productId == productId);
    cart.splice(foundCartItemIdx, 1);
  };

  const getCartItemAmount = productId => {
    const foundCartItem = findCartItemByProductId(productId);
    if (foundCartItem) {
      return (foundCartItem.price * foundCartItem.quantity).toFixed(2);
    }
    return 0;
  };

  const cartTotalAmount = computed(() => {
    const total = cart.reduce((acc, o) => {
      return acc + o.price * o.quantity;
    }, 0);

    return total.toFixed(2);
  });

  const cartTotalQuantity = computed(() => {
    const total = cart.reduce((acc, o) => {
      return acc + o.quantity;
    }, 0);

    return total;
  });

  return {
    cart,
    clearCart,
    addProductToCart,
    changeCartItemQuantity,
    removeCartItem,
    getCartItemAmount,
    cartTotalAmount,
    cartTotalQuantity
  };
});
