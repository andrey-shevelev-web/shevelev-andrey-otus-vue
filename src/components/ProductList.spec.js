import { describe, it, expect, vi, beforeEach } from 'vitest';
import { mount } from '@vue/test-utils';

import { setActivePinia, createPinia } from 'pinia';
import { jsonProducts } from '@/services/JsonProducts';
import * as ProductService from '@/services/ProductService';
import { useProductStore } from '@/stores/ProductStore';
import component from '@/components/ProductList.vue';

vi.mock('@/services/ProductService');
vi.mock('primevue/usetoast');

describe('ProductList', () => {
  beforeEach(async () => {
    const mockProducts = JSON.parse(jsonProducts);
    setActivePinia(createPinia());

    ProductService.getProducts = vi.fn().mockResolvedValue(mockProducts);
    const store = useProductStore();
    await store.loadProducts();
  });

  it('mounts without errors', async () => {
    const wrapper = mount(component, {
      global: {
        stubs: ['toast']
      }
    });
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
