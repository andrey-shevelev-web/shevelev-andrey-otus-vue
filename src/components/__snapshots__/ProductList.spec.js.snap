// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductList > mounts without error 1`] = `
"<div class="surface-ground px-4 py-8 md:px-6 lg:px-8">
  <div class="text-900 font-bold text-6xl mb-4 text-center">Products</div>
  <div class="p-toolbar p-component" role="toolbar" data-pc-name="toolbar" data-pc-section="root">
    <div class="p-toolbar-group-start p-toolbar-group-left" data-pc-section="start"><button class="p-button p-component" type="button" aria-label="Create Product" data-pc-name="button" data-pc-section="root"><span class="p-button-icon p-button-icon-left pi pi-plus" data-pc-section="icon"></span><span class="p-button-label" data-pc-section="label">Create Product</span>
        <!---->
      </button></div>
    <div class="p-toolbar-group-center" data-pc-section="center"></div>
    <div class="p-toolbar-group-end p-toolbar-group-right" data-pc-section="end"></div>
  </div>
  <!--v-if-->
  <div class="grid"></div>
</div>
<toast-stub></toast-stub>"
`;
